File: src/sys_switch/platforms/linux.py
from __future__ import annotations
import re
from typing import List, Optional

from .common import run, which, is_admin
from sys_switch.models import BootEntry


class LinuxBootManager:
    def __init__(self) -> None:
        self.efibootmgr = which('efibootmgr')
        self.grub_reboot = which('grub-reboot')
        self.grub_set_default = which('grub-set-default')

    def available(self) -> bool:
        return self.efibootmgr is not None or self.grub_reboot is not None

    def list_entries(self) -> List[BootEntry]:
        entries: List[BootEntry] = []
        if self.efibootmgr:
            cp = run([self.efibootmgr])
            text = cp.stdout
            # Parse BootCurrent and BootNext
            current = re.search(r"BootCurrent:\s*(\w+)", text)
            next_ = re.search(r"BootNext:\s*(\w+)", text)
            cur = current.group(1) if current else None
            nxt = next_.group(1) if next_ else None
            for m in re.finditer(r"Boot(\w+)\*?\s+(.+)", text):
                bid, desc = m.group(1), m.group(2).strip()
                entries.append(BootEntry(id=bid, description=desc, is_current=(bid==cur), is_next=(bid==nxt), extra=None))
            return entries
        # Fallback grub: list from /boot/grub/grub.cfg is complex; we provide minimal stub
        if self.grub_reboot:
            # Try grub-editenv list to get saved_entry
            ge = which('grub-editenv')
            if ge:
                cp = run([ge, 'list'])
                text = cp.stdout
                m = re.search(r"saved_entry=(.+)", text)
                saved = m.group(1) if m else None
                # Not enumerating actual menuentries; provide generic entries
                return [
                    BootEntry(id='0', description='GRUB default entry', is_current=False, is_next=(saved=='0'))
                ]
        return entries

    def set_next(self, entry_id: str) -> tuple[bool, str]:
        # Prefer efibootmgr BootNext
        if self.efibootmgr:
            if not is_admin():
                return False, '需要root权限运行 efibootmgr 才能设置 BootNext'
            cp = run([self.efibootmgr, '-n', entry_id])
            if cp.returncode == 0:
                return True, '已设置下次启动项: ' + entry_id
            return False, cp.stderr or cp.stdout
        # Fallback grub-reboot
        if self.grub_reboot:
            if not is_admin():
                return False, '需要root权限运行 grub-reboot 才能设置下次启动项'
            cp = run([self.grub_reboot, entry_id])
            if cp.returncode == 0:
                return True, '已设置 GRUB 下次启动项: ' + entry_id
            return False, cp.stderr or cp.stdout
        return False, '未找到可用的引导管理工具 (efibootmgr/grub-reboot)'

    def reboot_now(self) -> tuple[bool, str]:
        if not is_admin():
            return False, '需要root权限才能重启系统'
        cp = run(['systemctl', 'reboot'])
        return (cp.returncode == 0, cp.stderr or cp.stdout)

File: src/platform/linux.rs
//! Linux boot manager implementation using efibootmgr and grub-reboot.

use anyhow::{anyhow, Result};
use regex::Regex;
use std::process::Command;

use super::{BootEntry, BootManager};

/// Linux boot manager that uses efibootmgr or grub-reboot.
pub struct LinuxBootManager {
    efibootmgr_path: Option<String>,
    grub_reboot_path: Option<String>,
}

impl LinuxBootManager {
    /// Create a new Linux boot manager, detecting available tools.
    pub fn new() -> Self {
        Self {
            efibootmgr_path: which("efibootmgr"),
            grub_reboot_path: which("grub-reboot"),
        }
    }

    /// Check if running as root.
    fn is_root() -> bool {
        unsafe { libc::geteuid() == 0 }
    }
}

impl Default for LinuxBootManager {
    fn default() -> Self {
        Self::new()
    }
}

impl BootManager for LinuxBootManager {
    fn available(&self) -> bool {
        self.efibootmgr_path.is_some() || self.grub_reboot_path.is_some()
    }

    fn list_entries(&self) -> Result<Vec<BootEntry>> {
        let mut entries = Vec::new();

        if let Some(ref efibootmgr) = self.efibootmgr_path {
            let output = Command::new(efibootmgr).output()?;
            let text = String::from_utf8_lossy(&output.stdout);

            // Parse BootCurrent and BootNext
            let current_re = Regex::new(r"BootCurrent:\s*(\w+)")?;
            let next_re = Regex::new(r"BootNext:\s*(\w+)")?;
            let entry_re = Regex::new(r"Boot(\w+)\*?\s+(.+)")?;

            let current = current_re
                .captures(&text)
                .map(|c| c.get(1).unwrap().as_str().to_string());
            let next = next_re
                .captures(&text)
                .map(|c| c.get(1).unwrap().as_str().to_string());

            for cap in entry_re.captures_iter(&text) {
                let id = cap.get(1).unwrap().as_str().to_string();
                let desc = cap.get(2).unwrap().as_str().trim().to_string();
                let is_current = current.as_ref().map_or(false, |c| c == &id);
                let is_next = next.as_ref().map_or(false, |n| n == &id);

                entries.push(
                    BootEntry::new(id, desc)
                        .with_current(is_current)
                        .with_next(is_next),
                );
            }
            return Ok(entries);
        }

        // Fallback to grub
        if let Some(ref _grub_reboot) = self.grub_reboot_path {
            if let Some(grub_editenv) = which("grub-editenv") {
                let output = Command::new(grub_editenv).arg("list").output()?;
                let text = String::from_utf8_lossy(&output.stdout);

                let saved_re = Regex::new(r"saved_entry=(.+)")?;
                let saved = saved_re
                    .captures(&text)
                    .map(|c| c.get(1).unwrap().as_str().to_string());

                entries.push(
                    BootEntry::new("0", "GRUB default entry")
                        .with_next(saved.as_ref().map_or(false, |s| s == "0")),
                );
            }
        }

        Ok(entries)
    }

    fn set_next(&self, entry_id: &str) -> Result<String> {
        if !Self::is_root() {
            return Err(anyhow!("Root privileges required"));
        }

        // Prefer efibootmgr
        if let Some(ref efibootmgr) = self.efibootmgr_path {
            let output = Command::new(efibootmgr).args(["-n", entry_id]).output()?;

            if output.status.success() {
                return Ok(format!("Next boot entry set to: {}", entry_id));
            }
            let stderr = String::from_utf8_lossy(&output.stderr);
            return Err(anyhow!("{}", stderr));
        }

        // Fallback to grub-reboot
        if let Some(ref grub_reboot) = self.grub_reboot_path {
            let output = Command::new(grub_reboot).arg(entry_id).output()?;

            if output.status.success() {
                return Ok(format!("GRUB next boot entry set to: {}", entry_id));
            }
            let stderr = String::from_utf8_lossy(&output.stderr);
            return Err(anyhow!("{}", stderr));
        }

        Err(anyhow!(
            "No supported boot manager found (efibootmgr/grub-reboot)"
        ))
    }

    fn reboot_now(&self) -> Result<()> {
        if !Self::is_root() {
            return Err(anyhow!("Root privileges required to reboot"));
        }

        let output = Command::new("systemctl").arg("reboot").output()?;

        if output.status.success() {
            Ok(())
        } else {
            let stderr = String::from_utf8_lossy(&output.stderr);
            Err(anyhow!("{}", stderr))
        }
    }
}

/// Find an executable in PATH.
fn which(cmd: &str) -> Option<String> {
    std::env::var_os("PATH").and_then(|paths| {
        std::env::split_paths(&paths)
            .filter_map(|dir| {
                let full_path = dir.join(cmd);
                if full_path.is_file() {
                    Some(full_path.to_string_lossy().into_owned())
                } else {
                    None
                }
            })
            .next()
    })
}
